### PR TITLE
doc, feat: added description of deutsch_jozsa algorithm #2

### PR DESCRIPTION
**Pull request by AmandaStromdahl**
Added a description of deutsch_jozsa(), from the set of quantum algorithms.

### CODE REVIEW

File: complex_functions/deutsch_jozsa.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import numpy as np
```
Comment (bastienfaivre) [inline complex_functions/deutsch_jozsa.py:1]: Can you add the small comment saying that this function is not ours and belongs to the Python repo. See my files for the comment line

File: report.md [hunk 2]
```diff
@@ -29,36 +29,47 @@ The eight functions are presented in the table below. Every function has a CCN o
 
 | Function                                                            | Location                         | CCN, Lizard | CCN, manual (Student 1) | CCN, manual (Student 2) | NLOC, Lizard |
 | ------------------------------------------------------------------- | -------------------------------- | ----------- | ----------------------- | ----------------------- | ------------ |
-| Dijkstra(graph, V, src)                                             | graphs/dijkstra_2.py             | 8           | -                       | -                       | 15           |
+| dj_oracle(case, num_qubits)                                         | 9                                | 9           | -                       | 25                      |
```
Comment (bastienfaivre) [inline report.md]: I found the same CCN, you can add 9 to "manual student 2"

File: report.md [hunk 3]
```diff
@@ -29,36 +29,47 @@ The eight functions are presented in the table below. Every function has a CCN o
...
+1. **Are the functions just complex, or also long?**
+2. **What is the purpose of the functions?**
+3. **Are exceptions taken into account in the given measurements?**
+4. **Is the documentation clear w.r.t. all the possible outcomes?**
+
+### <a id="dj_oracle"></a>Deutsch Jozsa ([file](quantum/deutsch_jozsa.py))
+
+The Deutsch-Jozsa problem features an oracle function that takes an n-bit input and outputs 0 or 1. The function is either constant (returns a constant value) or balanced (meaning it returns 0 half of the time and 1 the rest of the time). The purpose of the examined method is to create a Quantum Cirtcuit (a model) of an oracle function which is either balanced or constant depending on the parameter `case`, and that handles `num_qubits`-bit input values.
+
+The function handles multiple cases and needs to iterate through `num_qubits` when setting the output values for the model. At the same time, the only way to exit the function is to return the model. This is why the cyclomatic complexity is so high. The documentation is clear about the only possible outcome of the function.
```
Comment (bastienfaivre) [inline report.md]: Can you explain more precisely what are the possible outputs of the algorithm ?

**Review by bastienfaivre — COMMENTED** [review]
Nice job ! Please have a look on my comments

File: complex_functions/deutsch_jozsa.py (hunk 1 above)
Comment (AmandaStromdahl) [reply complex_functions/deutsch_jozsa.py:1]: Great suggestion, I'll add that.

File: report.md (hunk 2 above)
Comment (AmandaStromdahl) [reply report.md]: Yes, I'll add that too.

File: report.md (hunk 3 above)
Comment (AmandaStromdahl) [reply report.md]: Sure thing

**Comment by AmandaStromdahl** [comment]
I implemented all three suggestions.

**Comment by bastienfaivre** [comment]
LGTM

**Pull request merged**